src/charm.py

```diff
@@ -0,0 +1,256 @@
+#!/usr/bin/env python3
+
+import os
+import subprocess
+import socket
+import re
+import pwd
+import sys
+sys.path.append('lib')  # noqa
+
+from ops.charm import CharmBase, CharmEvents
+from ops.framework import EventBase, EventSource, StoredState
+from ops.main import main
+from ops.model import (
+    ActiveStatus,
+    BlockedStatus,
+    MaintenanceStatus,
+    WaitingStatus,
+    ModelError,
+)
+from cluster import CockroachDBCluster
+from interface_proxy_listen_tcp import ProxyListenTcpInterfaceProvides
+
+from jinja2 import Environment, FileSystemLoader
+from datetime import timedelta
+from time import sleep
+
+
+class CockroachStartedEvent(EventBase):
+    pass
+
+
+class ClusterInitializedEvent(EventBase):
+    def __init__(self, handle, cluster_id):
+        super().__init__(handle)
+        self.cluster_id = cluster_id
+
+    def snapshot(self):
+        return self.cluster_id
+
+    def restore(self, cluster_id):
+        self.cluster_id = cluster_id
+
+
+class CockroachDBCharmEvents(CharmEvents):
+    cockroachdb_started = EventSource(CockroachStartedEvent)
+    cluster_initialized = EventSource(ClusterInitializedEvent)
+
+
+class CockroachDBCharm(CharmBase):
+    on = CockroachDBCharmEvents()
+    state = StoredState()
+
+    COCKROACHDB_SERVICE = 'cockroachdb.service'
+    SYSTEMD_SERVICE_FILE = f'/etc/systemd/system/{COCKROACHDB_SERVICE}'
+    WORKING_DIRECTORY = '/var/lib/cockroach'
+    COCKROACH_INSTALL_DIR = '/usr/local/bin'
+    COCKROACH_BINARY_PATH = f'{COCKROACH_INSTALL_DIR}/cockroach'
+    COCKROACH_USERNAME = 'cockroach'
+    PSQL_PORT = 26257
+    HTTP_PORT = 8080
+
+    MAX_RETRIES = 10
+    RETRY_TIMEOUT = timedelta(milliseconds=125)
+
+    def __init__(self, framework, key):
+        super().__init__(framework, key)
+
+        self.state.set_default(is_started=False)
+
+        for event in (self.on.install,
+                      self.on.start,
+                      # self.on.upgrade_charm,
+                      self.on.config_changed,
+                      self.on.cluster_relation_changed,
+                      self.on.cockroachdb_started,
+                      self.on.proxy_listen_tcp_relation_joined):
+            self.framework.observe(event, self)
+
+        self.cluster = CockroachDBCluster(self, 'cluster')
+        self.tcp_load_balancer = ProxyListenTcpInterfaceProvides(self, 'proxy-listen-tcp')
```

Comment:
We need a better type name for that one.

Reply:
Agreed in https://github.com/canonical/cockroachdb-operator/pull/1#discussion_r402943185